src/Kestrel.Core/Internal/Http2/Http2Stream.FeatureCollection.cs

```diff
@@ -246,13 +246,13 @@ MinDataRate IHttpMinResponseDataRateFeature.MinDataRate
 
         object IFeatureCollection.this[Type key]
         {
-            get => FastFeatureGet(key);
```

Comment:
Will remote IP be a connection level feature? We mutate that all the time using x-forwarded-for. It should be reset per request as each request may have different forwarders.

So far I think you're handling the IPs correctly, copying them when you create the frame, but I'm less sure about the other features and their mutable fields.